File: By-Swift/By-Swift/Programmers/[2020-카카오-여름-인턴십]/보석쇼핑.swift
import Foundation

func solution(_ gems:[String]) -> [Int] {
    var cart = [String : Int]()
    var ans = [1,gems.count]
    let gemNum = Set(gems).count
    var start = 0
    
    for end in 0..<gems.count {
         cart[gems[end],default: 0] += 1
        
        while (cart.keys.count == gemNum) && (start <= end) {
            
            ans = (ans.last! - ans.first!) > (end-start) ? [start+1, end+1] : ans

            cart[gems[start],default: 0] -= 1
            if cart[gems[start]]! == 0 {
                cart.removeValue(forKey: gems[start])
            }
            start += 1
        }
    }
    return ans
}

File: By-Python/programmers/[2020-카카오-여름-인턴십]/보석쇼핑.py
from collections import defaultdict

# 통과 # len(set(cart))대신 dictionary사용
def solution(gems):
    if len(gems) == 1: return [1]
    
    gemNum = len(set(gems))
    cart = defaultdict(list)
    answer = [1,len(gems)]
    start = 0
    
    for end in range(len(gems)):
        cart[gems[end]] = cart.get(gems[end], 0) + 1
        
        while (len(cart) == gemNum) and (start <= end):
            if (answer[1]-answer[0]) > (end-start):
                answer = [start+1, end+1]
            
            cart[gems[start]] -= 1
            
            if cart[gems[start]] == 0:
                del cart[gems[start]]
            
            start += 1
                
    return answer